File: ocaml/space-age/space_age.ml
type planet =
  | Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Neptune
  | Uranus

let age_on planet sec_of_age =
  float_of_int sec_of_age /. 31557600.0
  /.
  match planet with
  | Mercury -> 0.2408467
  | Venus -> 0.61519726
  | Earth -> 1.
  | Mars -> 1.8808158
  | Jupiter -> 11.862615
  | Saturn -> 29.447498
  | Neptune -> 164.79132
  | Uranus -> 84.016846

File: elixir/space-age/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  defp s_to_y(seconds), do: seconds / 31_557_600

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(planet, seconds) when planet == :mercury, do: {:ok, s_to_y(seconds) / 0.2408467}
  def age_on(planet, seconds) when planet == :venus, do: {:ok, s_to_y(seconds) / 0.61519726}
  def age_on(planet, seconds) when planet == :earth, do: {:ok, s_to_y(seconds)}
  def age_on(planet, seconds) when planet == :mars, do: {:ok, s_to_y(seconds) / 1.8808158}
  def age_on(planet, seconds) when planet == :jupiter, do: {:ok, s_to_y(seconds) / 11.862615}
  def age_on(planet, seconds) when planet == :saturn, do: {:ok, s_to_y(seconds) / 29.447498}
  def age_on(planet, seconds) when planet == :uranus, do: {:ok, s_to_y(seconds) / 84.016846}
  def age_on(planet, seconds) when planet == :neptune, do: {:ok, s_to_y(seconds) / 164.79132}
  def age_on(_planet, _seconds), do: {:error, "not a planet"}
end
